packages/ai/core/util/call-chat-api-v2.ts

```diff
@@ -0,0 +1,110 @@
+import { IdGenerator, JSONValue, UseChatOptions } from '../types';
+import { ChatStore } from './chat-store';
+import { processChatResponseV2 } from './process-chat-response-v2';
+import { processChatTextResponseV2 } from './process-chat-text-response-v2';
+
+// use function to allow for mocking in tests:
+const getOriginalFetch = () => fetch;
+
+export async function callChatApiV2({
+  api,
+  body,
+  streamProtocol = 'data',
+  credentials,
+  headers,
+  abortController,
+  restoreMessagesOnFailure,
+  onResponse,
+  onUpdate,
+  onFinish,
+  onToolCall,
+  generateId,
+  fetch = getOriginalFetch(),
+  store,
+  chatId,
+}: {
+  api: string;
+  body: Record<string, any>;
+  streamProtocol: 'data' | 'text' | undefined;
+  credentials: RequestCredentials | undefined;
+  headers: HeadersInit | undefined;
+  abortController: (() => AbortController | null) | undefined;
+  restoreMessagesOnFailure: () => void;
+  onResponse: ((response: Response) => void | Promise<void>) | undefined;
+  onUpdate: (options: { data: JSONValue[] | undefined }) => void;
+  onFinish: UseChatOptions['onFinish'];
+  onToolCall: UseChatOptions['onToolCall'];
+  generateId: IdGenerator;
+  fetch: ReturnType<typeof getOriginalFetch> | undefined;
+  store: ChatStore;
+  chatId: string;
+}) {
+  const response = await fetch(api, {
+    method: 'POST',
+    body: JSON.stringify(body),
+    headers: {
+      'Content-Type': 'application/json',
+      ...headers,
+    },
+    signal: abortController?.()?.signal,
+    credentials,
+  }).catch(err => {
+    restoreMessagesOnFailure();
+    throw err;
+  });
+
+  if (onResponse) {
+    try {
+      await onResponse(response);
+    } catch (err) {
+      throw err;
+    }
```

Comment:
why is the catch-rethrow required?

Reply:
I'm not too sure! This is a relic from the original callChatApi, can dig though to understand if it's necessary!